contracts/Staking.sol

```diff
@@ -0,0 +1,466 @@
+  }
+
+  function _handleRedelegateAckPackage(RLPDecode.Iterator memory iter) internal {
+    RedelegateAckPackage memory pack;
+
+    bool success = false;
+    uint256 idx = 0;
+    while (iter.hasNext()) {
+      if (idx == 0) {
+        pack.delegator = address(uint160(iter.next().toAddress()));
+      } else if (idx == 1) {
+        pack.valSrc = address(uint160(iter.next().toAddress()));
+      } else if (idx == 2) {
+        pack.valDst = address(uint160(iter.next().toAddress()));
+      } else if (idx == 3) {
+        pack.amount = uint256(iter.next().toUint());
+      } else if (idx == 4) {
+        pack.errCode = uint8(iter.next().toUint());
+        success = true;
+      } else {
+        break;
+      }
+      idx++;
+    }
+    require(success, "rlp decode ack package failed");
+
+    delegatedOfValidator[pack.delegator][pack.valSrc] = delegatedOfValidator[pack.delegator][pack.valSrc].add(pack.amount);
+    delegatedOfValidator[pack.delegator][pack.valDst] = delegatedOfValidator[pack.delegator][pack.valDst].sub(pack.amount);
+    pendingRedelegate[pack.delegator][pack.valSrc][pack.valDst] = 0;
+
+    emit failedRedelegate(pack.delegator, pack.valSrc, pack.valDst, pack.amount, pack.errCode);
+  }
+
+  function _handleDistributeRewardSynPackage(RLPDecode.Iterator memory iter) internal returns(uint32) {
+    DistributeRewardSynPackage memory pack;
+
+    bool success = false;
+    uint256 idx = 0;
+    while (iter.hasNext()) {
+      if (idx == 0) {
+        pack.amount = uint256(iter.next().toUint());
+      } else if (idx == 1) {
+        pack.recipient = address(uint160(iter.next().toAddress()));
+        success = true;
+      } else {
+        break;
+      }
+      idx++;
+    }
+    if (!success) {
+      return ERROR_FAIL_DECODE;
+    }
+
+    bool ok = ITokenHub(TOKEN_HUB_ADDR).withdrawStakingBNB(pack.amount);
+    if (!ok) {
+      return ERROR_WITHDRAW_BNB;
+    }
+
+    distributedReward[pack.recipient] = distributedReward[pack.recipient].add(pack.amount);
+    emit rewardReceived(pack.recipient, pack.amount);
+
+    return CODE_OK;
+  }
+
+  function _handleDistributeUndelegatedSynPackage(RLPDecode.Iterator memory iter) internal returns(uint32) {
+    DistributeUndelegatedSynPackage memory pack;
+
+    bool success = false;
+    uint256 idx = 0;
+    while (iter.hasNext()) {
+      if (idx == 0) {
+        pack.amount = uint256(iter.next().toUint());
+      } else if (idx == 1) {
+        pack.recipient = address(uint160(iter.next().toAddress()));
+      } else if (idx == 2) {
+        pack.validator = address(uint160(iter.next().toAddress()));
+        success = true;
+      } else {
+        break;
+      }
+      idx++;
+    }
+    if (!success) {
+      return ERROR_FAIL_DECODE;
+    }
+
+    bool ok = ITokenHub(TOKEN_HUB_ADDR).withdrawStakingBNB(pack.amount);
+    if (!ok) {
+      return ERROR_WITHDRAW_BNB;
+    }
+
+    pendingUndelegated[pack.recipient][pack.validator] = 0;
```

Comment:
I am worried about assigning it to zero directly, shall we just do minus?

Reply:
Fixed